lib/hedwig/responder.ex

```diff
@@ -193,48 +176,74 @@ defmodule Hedwig.Responder do
   end
 
   defmacro __before_compile__(_env) do
-    quote do
-      @doc false
-      def usage(name) do
-        @usage
-        |> Enum.map(&String.strip/1)
-        |> Enum.map(&(String.replace(&1, "hedwig", name)))
+    quote location: :keep do
+
+      def init({aka, name, opts, robot}) do
+        :ok = GenServer.cast(self(), :compile_responders)
+
+        {:ok, %{
+          aka: aka,
+          name: name,
+          opts: opts,
+          responders: [],
+          robot: robot}}
       end
 
-      def __hearers__ do
-        @hear
+      def handle_cast(:compile_responders, %{aka: aka, name: name} = state) do
+        {:noreply, %{state | responders: compile_responders(name, aka)}}
       end
 
-      def __responders__ do
-        @respond
+      def handle_cast({:dispatch, msg}, state) do
+        {:noreply, dispatch_responders(msg, state)}
       end
 
-      @doc false
-      def install(robot, opts) do
-        hearers =
-          __hearers__
-          |> Enum.map(&install_hearer(&1, robot, opts))
-          |> Enum.map(&Task.await/1)
+      def dispatch_responders(msg, %{responders: responders} = state) do
+        Enum.reduce responders, state, fn responder, state ->
+          case dispatch_responder(responder, msg, state) do
+            :ok ->
+              state
+            {:ok, state} ->
+              state
+          end
+        end
+      end
 
-        responders =
-          __responders__
-          |> Enum.map(&install_responder(&1, robot, opts))
-          |> Enum.map(&Task.await/1)
+      def dispatch_responder({regex, fun}, %{text: text} = msg, state) do
+        if Regex.match?(regex, text) do
+          msg = %{msg | matches: find_matches(regex, text)}
+          apply(__MODULE__, fun, [msg, state])
+        else
+          :ok
+        end
+      end
 
-        List.flatten([hearers, responders])
+      defp find_matches(regex, text) do
+        case Regex.names(regex) do
+          []  ->
+            matches = Regex.run(regex, text)
+            Enum.reduce(Enum.with_index(matches), %{}, fn {match, index}, acc ->
+              Map.put(acc, index, match)
+            end)
+          _ ->
+            Regex.named_captures(regex, text)
+        end
       end
 
-      defp install_hearer({regex, fun}, _robot, opts) do
-        Task.async(fn ->
-          {regex, __MODULE__, fun, Enum.into(opts, %{})}
-        end)
+      @doc false
+      def usage(name) do
+        @usage
+        |> Enum.map(&String.strip/1)
```

Comment:
What about `&(String.strip(&1) |> String.replace("hedwig", name))`...or even:

```elixir
def usage(name) do
  import String
  @usage
  |> Enum.map(&(strip(&1) |> replace("hedwig", name))
```

Or is this worse? :)